### ISSUE TITLE
Do not inject setState/ref into dependencies

### ISSUE THREAD

**Issue by Glinkis** [yuchi/hooks.macro#10]
Currently testing this macro, so far with great success!

_But_ I have one minor nitpick.

Currently this:
```js
const [state, setState] = useState(0)
useAutoEffect(() => {
   setState(1)
})
```
Is converted to:
```js
const [state, setState] = useState(0)
useEffect(() => {
   setState(1)
}, [setState])
```

But adding `setState` to the dependency array is redundant, since react guarantees it to be referentially stable. The same is true for `useRef`.

The `react-hooks/exhaustive-deps` eslint takes both of them into consideration.

**Comment by yuchi** [comment]
This is a very nice suggestion! We could also exclude:

1. the state value if the setter is never used
2. memoized values/callbacks that use an empty dependencies array

**Comment by Glinkis** [comment]
Thanks for the response!
I'm not sure I quite understand your suggestions. It would be awesome if you could show an example of each of them.

**Comment by yuchi** [comment]
```js
function MyComponent() {
  const [id] = React.useState(() => Math.random());

  // Doesn’t require [id] since it will never change
  useAutoEffect(() => {
    console.log(id);
  });

  const [counter, setCounter] = React.useState(0);

  // Again, `setCounter` is not used, `counter` will never change!
  const values = useMemo(new Array(counter));

  const handleLog = React.useCallback(() => {}, []);

  // `handleLog` will (*should*) never change, using as a dependency is reduntant
  useAutoEffect(() => handleLog('Hallo!'));
}
```

**Comment by yuchi** [comment]
Found the relative code in the ESLint exhaustive-deps rules: https://github.com/facebook/react/blob/fe2cb525542443aaf1447c4069354c12659fd186/packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js#L296

Should be straight forward to add them to this macro.

**Comment by yuchi** [comment]
Please see #11 and #12 for further discussion on those additional optimizations.

**Comment by yuchi** [comment]
@Glinkis you can see this implemented in v1.1.0. Should bring those small perf improvements :)

**Comment by Glinkis** [comment]
Thanks!
